Reduce colrefs by skipping unused ones

asubramanya opened the pull request:
Currently we retain all the colrefs of a table even though we only reference a
few of them in the original query. This patch marks the colrefs as used if
either its a system column or if its referenced in the query and unused
otherwise. When we derive the relational properties, we exclude the unused
columns from the output columns list. For a SELECT COUNT(*) query on a 1600
column table which has a 10 way join, we see an improvement of approximately
40%.

Co-authored-by: Abhijit Subramanya <asubramanya@pivotal.io>

zellerh commented on libgpopt/include/gpopt/translate/CTranslatorDXLToExpr.h:
```diff
@@ -398,6 +398,9 @@ namespace gpopt
 
 		public:
+
+			static BOOL translating;
```
The public member works just fine, but to be more conventional, maybe make this a private member and define methods to set/reset this?

hardikar replied:
Why does this need to be static when the class is used non-static-ly. That is, we create an object of CTranslatorDXLToExpr when we need to use it. Using a static boolean variable to keep track of this could create trouble if there are multiple translations happening at the same time (doesn't happen right now). Consider rewriting this as a member variable.

zellerh commented on libgpopt/src/operators/CLogicalGet.cpp:
```diff
@@ -231,7 +232,13 @@ CLogicalGet::PcrsDeriveOutput
-	pcrs->Include(m_pdrgpcrOutput);
+	for (ULONG i = 0; i < m_pdrgpcrOutput->Size(); i++)
+	{
+		if ((*m_pdrgpcrOutput)[i]->IsUsed() || CTranslatorDXLToExpr::translating)
```
We should add a comment here why we have this extra check.

zellerh commented on server/src/unittest/gpopt/operators/CExpressionTest.cpp:
```diff
@@ -180,6 +181,28 @@ CExpressionTest::EresUnittest_SimpleOps()
+			colrefs = popGet->PdrgpcrOutput();
+		}
+
+		if (NULL != popDynGet)
```
Use an else if here?

hardikar replied:
Hmm.. I'd prefer to use an if case here comparing `pexpr->Pop()->Eopid()`, instead of dynamic_casts

bhuvnesh2703 commented on data/dxl/minidump/AnySubq-With-NonScalarSubqueryChild-1.mdp:
```diff
@@ -541,48 +541,40 @@
                   <dxl:GroupingColumns>
-                    <dxl:GroupingColumn ColId="10"/>
                     <dxl:GroupingColumn ColId="11"/>
-                    <dxl:GroupingColumn ColId="12"/>
```
2 grouping cols removed? could you explain why?

bhuvnesh2703 commented on data/dxl/minidump/AssertMaxOneRow.mdp:
```diff
@@ -1783,42 +1725,14 @@ where oid= (select reltoastrelid from pg_class where relname='fsts_alter_set_sto
                       <dxl:Comparison ComparisonOperator="=" OperatorMdid="0.93.1.0">
                         <dxl:Ident ColId="76" ColName="relname" TypeMdid="0.19.1.0"/>
-                        <dxl:ConstValue TypeMdid="0.19.1.0" IsNull="false" IsByValue="false" Value="ZnN0c19hbHRlcl9zZXRfc3RvcmFnZV90YWJsZQAAAAAAAAAAAAAAAAAAAAAA&#10;AAAAAAAAAAAAAAAAAAAAAAAAAA=="/>
+                        <dxl:ConstValue TypeMdid="0.19.1.0" IsNull="false" IsByValue="false" Value="ZnN0c19hbHRlcl9zZXRfc3RvcmFnZV90YWJsZQAAAAAAAAAAAAAAAAAAAAAA&#xA;AAAAAAAAAAAAAAAAAAAAAAAAAA=="/>
```
why the change from `#10` => `xA`

bhuvnesh2703 commented on data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnionAllWithNoRedistributableColumns.mdp:
```diff
@@ -246,130 +246,58 @@ EXPLAIN SELECT xmin FROM tbl1 UNION ALL SELECT xmin FROM tbl2;
           </dxl:ProjList>
           <dxl:Filter/>
-          <dxl:Result>
+          <dxl:TableScan>
```
this test is for parallel union all, so we should see a redistribute motion (no-op) on top of every child of append? why its removed?

bhuvnesh2703 commented on libgpopt/include/gpopt/base/CUtils.h:
```diff
@@ -1194,7 +1194,6 @@ namespace gpopt
 		GPOS_ASSERT(NULL != pexprLeft);
 		GPOS_ASSERT(NULL != pexprRight);
-		GPOS_ASSERT(NULL != pcrInner);
 
```
we expect inner to be NULL now sometimes, why?

hardikar replied:
@asubramanya Was this concern addressed?

asubramanya replied:
Seems like this change is no longer needed now. So reverted it back to the old code.

bhuvnesh2703 commented on libgpopt/src/base/CColumnFactory.cpp:
```diff
@@ -262,12 +262,18 @@ CColumnFactory::PcrCreate
 	ULONG id = m_aul.Incr();
 	
-	return PcrCreate(pcoldesc, id, name, ulOpSource);
+	CColRef *colref = PcrCreate(pcoldesc, id, name, ulOpSource);
```
minor: what if we pass mark_as_used as parameter to PcrCreate?

bhuvnesh2703 commented on libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -7302,20 +7302,24 @@ CTranslatorExprToDXL::MakeDXLTableDescr
-		// use the col ref id for the corresponding output output column as 
+		CMDName *pmdnameCol = GPOS_NEW(m_mp) CMDName(m_mp, pcd->Name().Pstr());
+
+		// use the col ref id for the corresponding output output column as
```
`output output`

bhuvnesh2703 commented:
Few questions:
a) Instead of true / false for identifying if a column is used or not, could we have an enum which has the following states (chose any name)
1. Default State
2. Used
3. Unused

You can change Default State to Unused, but if something is marked Used we can't change it to Unused.
Do you forsee issue where we have something marked as used (really), but for some reason we mark it unused later because we don't look it up?

b) How we handle computed columns since they will not be originally in the query? Since we initialize starting with true, is it fine because of that?

sambitesh commented on libgpopt/src/optimizer/COptimizer.cpp:
```diff
@@ -243,8 +243,10 @@ COptimizer::PdxlnOptimize
 
 			// translate DXL Tree -> Expr Tree
 			CTranslatorDXLToExpr dxltr(mp, md_accessor);
+			CTranslatorDXLToExpr::translating = true;
```
Talked to @zellerh about this. It will be good to have comment added here.

hardikar commented on server/src/unittest/gpopt/translate/CTranslatorDXLToExprTest.cpp:
```diff
@@ -118,14 +118,14 @@ CTranslatorDXLToExprTest::Pexpr
 	// translate DXL Tree -> Expr Tree
 	CTranslatorDXLToExpr *pdxltr = GPOS_NEW(mp) CTranslatorDXLToExpr(mp, md_accessor);
-
+	pdxltr->translating = true;
```
Please move this inside PexprTranslateQuery. You can have it as a private variable then. I'm not sure why this is needed at all? Aren't you by-default translating if you're in the PexprTranslateQuery() function?

hardikar commented on libgpopt/src/translate/CTranslatorDXLToExprUtils.cpp:
```diff
@@ -258,7 +258,8 @@ CTranslatorDXLToExprUtils::Pdrgpcr
 	{
 		ULONG *pulColId = (*colids)[ul];
-		const CColRef *colref = colref_mapping->Find(pulColId);
+		CColRef *colref = colref_mapping->Find(pulColId);
```
Maybe replace these two lines with `LookupColRef(colref_mapping, pulColId) so that MarkAsUsed gets called automatically.

hardikar replied:
This still needs to be done?

hardikar replied:
I tried doing this - looks like it's going to be more work than I anticipated. It can be done later. Ignore it :)

hardikar commented on libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -7302,20 +7302,24 @@ CTranslatorExprToDXL::MakeDXLTableDescr
 		if (NULL != pdrgpcrOutput)
 		{
 			colref = (*pdrgpcrOutput)[ul];
+			if (!colref->IsUsed())
```
```suggestion
			// Don't send unused colrefs to improve performance
			if (!colref->IsUsed())
```

hardikar commented:
Ok i just saw @bhuvnesh2703 's commit, it seems he has the same suggestion(s).

ashuka24 commented on server/src/unittest/gpopt/minidump/CMiniDumperDXLTest.cpp:
```diff
@@ -151,7 +151,6 @@ CMiniDumperDXLTest::EresUnittest_Basic()
 													ptroutput->GetOutputColumnsDXLArray(),
 													ptroutput->GetCTEProducerDXLArray()
 													);
-		
```
this file change seems unnecessary

ashuka24 commented on server/src/unittest/CTestUtils.cpp:
```diff
@@ -3095,7 +3106,6 @@ CTestUtils::PexprReadQuery
 						ptroutput->GetOutputColumnsDXLArray(),
 						ptroutput->GetCTEProducerDXLArray()
 						);
-
```
unnecessary change

ashuka24 commented on server/src/unittest/gpopt/translate/CTranslatorDXLToExprTest.cpp:
```diff
@@ -118,14 +118,12 @@ CTranslatorDXLToExprTest::Pexpr
 
 	// translate DXL Tree -> Expr Tree
 	CTranslatorDXLToExpr *pdxltr = GPOS_NEW(mp) CTranslatorDXLToExpr(mp, md_accessor);
-
```
seems like unnecessary file change

asubramanya commented:
@hardikar @bhuvnesh2703 I changed the implementation to use the enums as you suggested and got rid of the translating flag. Also added a traceflag to enable/disable the feature.

hardikar commented on libnaucrates/include/naucrates/traceflags/traceflags.h:
```diff
@@ -177,6 +177,10 @@ namespace gpos
 		EopttraceEnableEagerAgg = 103030,
+		
+		// Enable reduction of unused colrefs
+		EopttraceTranslateUnusedColrefs = 103031,
```
// Disable the traceflag to remove unused colrefs

hardikar commented on libgpopt/src/operators/CLogical.cpp:
```diff
@@ -756,7 +756,7 @@ CLogical::PpcDeriveConstraintFromTable
 		pdrgpcrNonSystem->Append(colref);
 
-		if (pcoldesc->IsNullable())
+		if (pcoldesc->IsNullable() || !colref->IsUsed())
```
Should this be `!colref->IsUsed()` or `!colref->IsUnsed()`?

asubramanya replied:
Changed to check for Unused rather than not being Used.

hardikar commented on libgpopt/src/operators/CLogicalGet.cpp:
```diff
@@ -231,7 +232,17 @@ CLogicalGet::PcrsDeriveOutput
+		// We will know the entire list of columns which are referenced in the query only after
+		// translating the entire DXL to an expression. Hence we should not limit the output columns
+		// before we have processed the entire DXL.
+		if ((*m_pdrgpcrOutput)[i]->IsUsed() || (*m_pdrgpcrOutput)[i]->IsUnknown())
```
Alternatively, you could use colref->GetUsage() and return the enum. Return Used if it's a system column (also probably a good idea to set it correctly in the constructor). Honestly I'd much prefer this.
But, if you really want to use bool methods instead of just returning the enum, maybe s/IsUnknown/IsUsageUnknown/ or something like that?

asubramanya replied:
Sure. Its probably better to use the GetUsage() method.

asubramanya commented on data/dxl/minidump/AnySubq-With-NonScalarSubqueryChild-2.mdp:
```diff
@@ -607,7 +600,6 @@
                     <dxl:SortingColumnList>
                       <dxl:SortingColumn ColId="10" SortOperatorMdid="0.97.1.0" SortOperatorName="&lt;" SortNullsFirst="false"/>
                       <dxl:SortingColumn ColId="11" SortOperatorMdid="0.664.1.0" SortOperatorName="&lt;" SortNullsFirst="false"/>
-                      <dxl:SortingColumn ColId="12" SortOperatorMdid="0.97.1.0" SortOperatorName="&lt;" SortNullsFirst="false"/>
```
This column is not being referenced anywhere and its ok to remove it. The other columns are either system columns or distribution keys or they are referenced.

hardikar commented on libgpopt/src/base/CColumnFactory.cpp:
```diff
@@ -172,6 +172,7 @@ CColumnFactory::PcrCreate
 	// ensure uniqueness
 	GPOS_ASSERT(NULL == LookupColRef(id));
 	m_sht.Insert(colref);
+	colref->MarkAsUsed();
```
this comment is wrong. You can't define an interface based on where it's called. I'd rather there be no comment.

hardikar commented on libgpopt/src/translate/CTranslatorDXLToExpr.cpp:
```diff
@@ -58,6 +58,8 @@ using namespace gpmd;
 using namespace gpdxl;
 using namespace gpopt;
 
+//BOOL CTranslatorDXLToExpr::translating = false;
```
This still needs to be removed.
